### DISCUSSION TITLE
RBMSymm working?

### DISCUSSION THREAD

**Discussion by fabienalet** [netket/netket#838] [Q&A]
I am observing on several examples that `RBMSymm` is not working well specially for d>1 lattices or graphs, providing very bad variational energies.
From my previous experience with NetKet, symmetric machines were giving better results. I can imagine some situations where taking all symmetries may lead to slightly inferior results, but this is not what is happening here. I am suspecting a bug has been introduced somehow in the refactoring of `RBMSymm`. 

I may just use things wrongly (e.g. I'm confused by the fact that large alpha are now to be given) though I couldn't get any example or tutorial working properly with `RBMSymm`.

Here is an example on the 6x6  2d Heisenberg model (which shouldn't be too hard), comparing similar standard RBM (with low alpha=1), symmetric RBM to the exact result. The 4x4 can also display a similar issue but sometimes convergence to reasonable results is observed by tuning hyperparameters (learning rate, samples etc). 

Am I missing something obvious ?

```
import netket as nk
import numpy as np
import jax

L = 6
g = nk.graph.Hypercube(length=L, n_dim=2, pbc=True)

hi = nk.hilbert.Spin(s=1 / 2, N=g.n_nodes, total_sz=0)

ha = nk.operator.Heisenberg(hilbert=hi, graph=g)

# RBM Symmetric Machine
ma = nk.models.RBMSymm(
    symmetries=g.automorphisms(),
    alpha=24,
    use_visible_bias=False,
    use_hidden_bias=True,
    dtype=np.complex64,
)

sa = nk.sampler.MetropolisExchange(hi, graph=g, n_chains_per_rank=1)

vs = nk.vqs.MCState(sa, ma, n_samples=2000)
op = nk.optimizer.Sgd(learning_rate=0.01)
sr = nk.optimizer.SR(diag_shift=0.01)

vmc = nk.VMC(ha, op, variational_state=vs, preconditioner=sr)
print(f"RBM Symmetric # variational parameters: {vmc.state.n_parameters}")

for it in vmc.iter(200,10):
    print(it,vmc.energy)
e_RBM_Sym=vmc.energy.mean

# Now standard RBM, small alpha
ma = nk.models.RBM(
    alpha=1,
    use_visible_bias=False,
    use_hidden_bias=True,
    dtype=np.complex64,
)

vs2 = nk.vqs.MCState(sa, ma, n_samples=2000)
vmc2 = nk.VMC(ha, op, variational_state=vs2, preconditioner=sr)
print(f"#RBM  variational parameters: {vmc2.state.n_parameters}")

for it in vmc2.iter(200,10):
    print(it,vmc2.energy)

e_RBM_=vmc2.energy.mean

# Now exact
#import scipy
#exact_ens=scipy.sparse.linalg.eigsh(ha.to_sparse(),k=1,which='SA',return_eigenvectors=False)
exact_ens=-97.757568

print("E_exact=",exact_ens)
print("E_RBM=",vmc2.energy.mean)
print("E_RBM_symm",vmc.energy.mean)
```

which in the ends gives the following comparison
```
E_exact= -97.757568
E_RBM= (-96.77583496877267-0.04122783346235135j)
E_RBM_symm (-48.66666666666646+6.046244984027617e-14j)
```

The energy for the symmetric RBM can be different when changing a bit samples, number of iterations, but is most of the time completely off

**Comment by fabienalet** [comment]
Btw, I tried to isolate the problem on previous versions of v3 but couldn't pinpoint it. 
I could find it already starting from this commit
https://github.com/netket/netket/commit/15cf4f4ff00deb911d4a2720b4590c3c2d0beb3d
and perhaps earlier (but can't check because of some older mpi conflict on my machine)

**Reply by femtobit** [reply]
Hi Fabien, thanks for the detailed example. I haven't looked at in in detail yet, but two questions come to my mind:

1. Have you tried using [`g.translation_group()`](https://www.netket.org/docs/_generated/graph/netket.graph.Lattice.html#netket.graph.Lattice.translation_group) (or `g.translations()` is some of the previous beta versions) instead of `.automorphisms()` to check whether that makes a difference? (Using the larger group of all 6x6 space group symmetries is probably also the reason why a large alpha is required in your version.) While symmetries should make a state easier to learn, I do recall that using the full space or automorphism group (instead of the usually easier-to-handle translation group) can sometimes introduce convergence issues (which at least need some more careful tuning of optimization settings and probably initial params to overcome).
2. Is the `RBMSymm` stuck in an eigenstate at the `-48.66` energy you show above? That is, what is the variance in this state and is the MCMC sampling converged. (Printing `vmc2.energy`itself  should show all the relevant information already.)

**Reply by fabienalet** [reply]
Hi Damian,
thanks for the suggestions, I'll be checking this and let you know. 
I am aware that including all symmetries is not always beneficial, but I wouldn't expect it to be that bad. 
And of course, it could also be an exact eigenstate, I'll check it out.

**Reply by fabienalet** [reply]
It seems using only the translation group gives much better results for this example. Thanks a lot! I'll be checking on other cases where I observed this issue.
It would be great to have a tutorial / example where some (but not all) symmetries are included (e.g. translation + certain rotations, but not all).  Also similarly working examples on how to use in practice (e.g. specifying characters of various representations for standard lattices, to get access to ground-states in various symmetry sectors) the great work by @attila-i-szabo @chrisrothUT @femtobit would also be fantastic, and clearly beneficial to the community.

**Reply by femtobit** [reply]
Yes, an extended tutorial like this is definitely planned. We do have a demonstration of the GCNN architecture for a Honeycomb lattice with all automorphisms at https://www.netket.org/tutorials/G-CNN_Honeycomb.html.

Actually, thinking about that just now makes me wonder whether using a GCNN with a couple of layers (maybe already starting with two) could work better with the higher number of symmetries in your initial example. (@chrisrothUT You have more experience with GCNNs, do you think they could help here?)

**Reply by fabienalet** [reply]
Yes, I know this example, but I'm still unfamiliar with the full new group architecture to use it beyond standard lattices. For instance, how would you do (with GCNN or more simply RBMSymm) the same Honeycomb example but with the J1-J2 Heisenberg model (where J2 is for next nearest neighbors on the honeycomb) ? I know how to hard code it as a graph and use it in Netket, but the drawback is that I have to painfully implement by hand the symmetries on that specific graph...

**Reply by chrisrothUT** [reply]
@fabienalet I suspect that the issue is related to the initialization, namely the weights are too large. Right now the default initialization is Normal(sigma=0.1) which seems way too large for alpha=48.  

The standard deviation of the output will be sqrt(36 x 0.01 x 48 x 36) =  24. Based on this analysis you should set kernel_init = Normal(1/sqrt(36 x 48 x 36)). We should really fix this initialization. Another approach that would fix the issue is to use a GCNN with one layer.

> I know how to hard code it as a graph and use it in Netket, but the drawback is that I have to painfully implement by hand the symmetries on that specific graph...

You don't have to do this. You can just do something like `GCNN(symmetries=graph, num_layers=something, features=something)`. If your graph has a `point_group` it will use that, otherwise it will call `graph.automorphisms() ` to find the symmetries by hand. 

Unless your complaint is about the next neighbors, which is on the to-do list.

**Reply by fabienalet** [reply]
@chrisrothUT Not complaining at all ! :) Just eager to use all your nice improvements.
My first point was: can I use part of the point group (only some symmetries for instance)? This could be useful for some "tilted" clusters which don't have the full (i.e. thermodynamic limit) lattice symmetry -- e.g. a square lattice with 40 sites
And yes my second point is about next nearest neighbours, and more generally speaking: how can I still use (compatible) lattice symmetries with an arbitrary set of coupling constants ? It must be simple but I'm not sure how to proceed.

**Reply by chrisrothUT** [reply]
I think the simplest way to do this would be to create a new graph for the next neighbors.

```
nn_graph = nk.graph.Graph(edges=next_neighbor_edges)
ha = nk.operator.Heisenberg(..., graph=graph) + J2* nk.operator.Heisenberg(..., graph=nn_graph) 
```

Now if the new graph with the two different colors of edges has less symmetries that the previous graph, you can feed GCNN the point group manually. Let's say you have a rectangular lattice which only has a D_2 point group

`GCNN(..., symmetries=graph, point_group=nk.utils.group.planar.D(n=2))`

**Reply by fabienalet** [reply]
Thanks for the nice idea, I'll try it and see how far I can get

**Comment by gcarleo** [comment]
Yes, in my recent experience there is really an issue with the standard initialization, that I think was introduced in one of the latest commits? Can we restore the default initialization we had for this machine in 2.0?

**Reply by gcarleo** [reply]
Still, it is a bit strange that using all symmetries gives much worse results in this situation. Maybe it would be nice to see if including all the automorphisms and using smaller initializiations we converge to something reasonable!

**Reply by fabienalet** [reply]
I will give it a try ! 
Perhaps also on too small samples with a lot of symmetries, there are more chances to get stuck into an exact eigenstate as pointed out by @femtobit  (even though I don't think it was the case here). Btw, is there any trick in that case?

**Reply by chrisrothUT** [reply]
Yes the trick is to pre-optimize the phases, which @attila-i-szabo has written about. With GCNN, you just need to set the value `equal_amplitudes = True`. Then you need to load the parameters of the original model into a new model with  `equal_amplitudes = False`. So the pseudocode is something like

```
ma = GCNN(...., equal_amplitudes=True)
vstate = MCState(..., machine=ma)
gs = nk.driver.VMC(..., variational_state=vstate) 
gs.run()
saved_params = flax.core.unfreeze(vstate.parameters)

ma = GCNN(...., equal_amplitudes=False)
vstate = MCState(..., machine=ma)
vstate.parameters = saved_params
gs = nk.driver.VMC(..., variational_state=vstate) 
gs.run()

```

**Reply by PhilipVinc** [reply]
should we change the initializer?

**Reply by femtobit** [reply]
It'd probably be good to have a default that properly scales with network size.

**Reply by chrisrothUT** [reply]
Definitely. One easy fix is to have RBMSymm just call GCNN with 1 layer and cosh nonlinearity. Then it would be initialized properly and work with a character table. I can do that

**Reply by PhilipVinc** [reply]
I'd like to keep RBMSymm simple. Can't we just change the initializer of DenseSymm?
